File: 프로그래머스/lv2/42890. 후보키/후보키.java
import java.util.*;

class Solution {
    private ArrayList<ArrayList<Integer>> candidateKey = new ArrayList<>();
    private ArrayList<ArrayList<Integer>> keys = new ArrayList<ArrayList<Integer>>();
    
    private void combination(ArrayList<Integer> list, int[] numArr, int index, int r){
        if (r == 0) {
            keys.add(new ArrayList<>(list));
        } else if (r > 0) {
            for (int i = index; i < numArr.length; i++) {
                list.add(numArr[i]);
                combination(list, numArr, i + 1, r - 1);
                list.remove(list.size() - 1);
            }
        }
    }
    
    private void findKey(String[][] relation, int n) {
        int[] numArr = new int[relation[0].length];
        for (int i = 0; i < relation[0].length; i++) {
            numArr[i] = i;
        }
        combination(new ArrayList<Integer>(), numArr, 0, n);
        for (int i = 0; i < keys.size(); i++) {
            ArrayList<ArrayList<String>> temp = new ArrayList<>();
            ArrayList<Integer> key = keys.get(i);
            for (int j = 0; j < relation.length; j++) {
                ArrayList<String> inner = new ArrayList<>();
                for (int k = 0; k < key.size(); k++) {
                    if (!inner.contains(relation[j][key.get(k)])){
                        inner.add(relation[j][key.get(k)]); 
                    }
                }
                if (!temp.contains(inner)) {
                    temp.add(inner);
                }
            }
            
            if (temp.size() == relation.length) {
                boolean add = true;
                for (int j = 0; j < candidateKey.size(); j++) {
                    int count = 0;
                    ArrayList<Integer> cKey = candidateKey.get(j);
                    for (int k = 0; k < cKey.size(); k++) {
                        if (key.contains(cKey.get(k))) {
                            count++;
                        }
                    }
                    if (count == cKey.size()) {
                        add = false;
                    }
                }
                if (add) {
                    candidateKey.add(key);
                }
            }
        }
    } 
    
    public int solution(String[][] relation) {
        for (int j = 1; j <= relation[0].length; j++) {
            findKey(relation, j);
            keys.clear();
        }

        return candidateKey.size();
    }
}

File: 프로그래머스/lv2/42890. 후보키/후보키.py
from itertools import combinations as cb

candidate_key = []

def find_key(relation, n):
    a = [i for i in range(len(relation[0]))]
    keys = list(cb(a, n))
    for key in keys:
        temp = []
        for i in range(len(relation)):
            b = []
            for j in key:
                if relation[i][j] not in b:
                    b.append(relation[i][j])
            if b not in temp:
                temp.append(b)
        if len(temp) == len(relation):
            add = True
            for ckey in candidate_key:
                count = 0
                for k in range(len(ckey)):
                    if ckey[k] in key: 
                        count += 1
                if count == len(ckey):
                    add = False
            
            if add:
                candidate_key.append(key)


def solution(relation):
    for i in range(1, len(relation[0]) + 1):
        find_key(relation, i)
    return len(candidate_key)

